internal/envoy/v3/ratelimit.go

```diff
@@ -0,0 +1,61 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v3
+
+import (
+	envoy_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	envoy_config_filter_http_local_ratelimit_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/local_ratelimit/v3"
+	envoy_type_v3 "github.com/envoyproxy/go-control-plane/envoy/type/v3"
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/projectcontour/contour/internal/dag"
+	"github.com/projectcontour/contour/internal/protobuf"
+)
+
+// LocalRateLimitConfig returns a config for the HTTP local rate
+// limit filter.
+func LocalRateLimitConfig(config *dag.LocalRateLimitPolicy) *any.Any {
+	if config == nil {
+		return nil
+	}
+
+	c := &envoy_config_filter_http_local_ratelimit_v3.LocalRateLimit{
+		// TODO it'd be nice for this to be unique per vhost/route
```

Comment:
yup this would be super cool to be able to get stats per vshost/route.

Reply:
:+1: gotta figure out a naming scheme. Do you know off the top of your head if we do this anywhere else (vhost/route-specific stat prefix)?

Reply:
We have cluster & listener stats and just append the cluster name or vhost name to the stat. 

In the following example, the cluster is `projectcontour-marketing_info_80` which is `projectcontour-marketing` is the namespace, `info` is the svc name, and `80` is the listener it's tied to. 
```
envoy_cluster_upstream_cx_length_ms_bucket{envoy_cluster_name="projectcontour-marketing_info_80",le="50"}
```

So maybe we need something that's `vhost_route_%something%`. The tricky part is a route is the sum of the conditions which could also include a header.

Reply:
:+1: i'll see what i can figure out

Reply:
yeah, so if we want per-route stats, it'll have to be some big long ugly string that incorporates the prefix and any header matches. I think it could work, but it's a bit ugly.

Alternately, we could just roll everything up to the vhost, so only have a unique stats prefix per vhost. You lose some of the granular information, but it's maybe more easily parsed.

Reply:
I implemented a per-vhost prefix in https://github.com/projectcontour/contour/pull/3251/commits/021a0c664c4570a3f50963287031413214ece78e, so all routes in a given vhost will roll up to the vhost metrics. If we think having per-route metrics is important, I can work out a string representation of a route.